app/models/sms/decoder.rb

```diff
@@ -118,11 +126,41 @@ def find_form
 
   def find_user
     @user = @msg.user
-    log_authentication_failure unless @user && @user.active?
+    return if @user.present?
+
+    # get potential users
+    possible_users = User.by_phone(@msg.from).active
+    return log_authentication_failure unless possible_users.present?
+
+    # check auth code
+    auth_code = @tokens.first if @tokens.first =~ AUTH_CODE_FORMAT
+    @user ||= possible_users.find_by(sms_auth_code: auth_code)
+
+
+    # check assigned to mission
+    possible_users = possible_users.assigned_to(@msg.mission)
```

Comment:
Technically this line and line 136 should not be run if `@user` is already set. This line will fire an unnecessary query. Can you wrap these blocks in `if`s instead of using `||=`?